src/test/java/net/snowflake/client/jdbc/DefaultResultStreamProviderTest.java

```diff
@@ -0,0 +1,120 @@
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.apache.http.Header;
+import org.apache.http.HttpResponse;
+import org.junit.Before;
+import org.junit.Test;
+
+public class DefaultResultStreamProviderTest {
+
+  private DefaultResultStreamProvider resultStreamProvider;
+  private HttpResponse mockResponse;
+
+  @Before
+  public void setUp() {
+    resultStreamProvider = new DefaultResultStreamProvider();
+    mockResponse = mock(HttpResponse.class);
+  }
+
+  private InputStream invokeDetectContentEncodingAndGetInputStream(
+      HttpResponse response, InputStream inputStream) throws Exception {
+    Method method =
+        DefaultResultStreamProvider.class.getDeclaredMethod(
+            "detectContentEncodingAndGetInputStream", HttpResponse.class, InputStream.class);
+    method.setAccessible(true);
+    return (InputStream) method.invoke(resultStreamProvider, response, inputStream);
+  }
+
+  @Test
+  public void testDetectContentEncodingAndGetInputStream_Gzip() throws Exception {
+    // Mocking gzip content encoding
+    Header encodingHeader = mock(Header.class);
+    when(encodingHeader.getValue()).thenReturn("gzip");
+    when(mockResponse.getFirstHeader("Content-Encoding")).thenReturn(encodingHeader);
+
+    // Original data to compress and validate
+    String originalData = "Some data in GZIP";
+
+    // Creating a gzip byte array using GZIPOutputStream
+    byte[] gzipData;
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream)) {
+      gzipOutputStream.write(originalData.getBytes(StandardCharsets.UTF_8));
+      gzipOutputStream.close(); // close to flush and finish the compression
+      gzipData = byteArrayOutputStream.toByteArray();
+    }
+
+    // Mocking input stream with the gzip data
+    InputStream gzipStream = new ByteArrayInputStream(gzipData);
+
+    // Call the private method using reflection
+    InputStream resultStream =
+        invokeDetectContentEncodingAndGetInputStream(mockResponse, gzipStream);
+
+    // Decompress and validate the data matches original
+    ByteArrayOutputStream decompressedOutput = new ByteArrayOutputStream();
+    byte[] buffer = new byte[1024];
+    int bytesRead;
+    try (GZIPInputStream gzipInputStream = (GZIPInputStream) resultStream) {
+      while ((bytesRead = gzipInputStream.read(buffer)) != -1) {
+        decompressedOutput.write(buffer, 0, bytesRead);
+      }
+    }
+    String decompressedData = new String(decompressedOutput.toByteArray(), StandardCharsets.UTF_8);
+
+    assertEquals(originalData, decompressedData);
+  }
+
+  @Test
+  public void testDetectContentEncodingAndGetInputStream_Zstd() throws Exception {
```

Comment:
The problem with such approach is that if we compress and decompress using the same tool, which contains some bug, we will silently pass a test. But there is probably no better approach now unless we use a fixed file which is compressed somewhere else.

Reply:
Possibly we could use some system utility like command line `zstd` in a subprocess to have this 3rd party library for verification? We'd need to ensure it's available on test environments though

Reply:
I think for now we should wait for full backend support of ZSTD chunks compression and then properly test the integration, because the most important thing is whether we are able to decompress what backend compresses. 3rd party tools are of smaller importance here IMO